File: Trees/145. Binary Tree Postorder Traversal/Solution.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # N : no of Nodes
        # Time Complexity: O(N)
        # Space Complexity: O(2N)
        # postOrder = []
        # def dfs(root):
        #     if root is None:
        #         return
        #     dfs(root.left)
        #     dfs(root.right)
        #     postOrder.append(root.val)
        #     return
        # dfs(root)
        # return postOrder
        

        # N : no of Nodes
        # Time Complexity: O(N)
        # Space Complexity: O(2N)
        postOrder = []
        stack = [root]
        while stack:
            node = stack.pop()
            if node is None:
                continue
            postOrder.append(node.val)
            stack.append(node.left)
            stack.append(node.right)
        return postOrder[::-1]

File: Trees/145. Binary Tree Postorder Traversal/Solution.java
import java.util.*;

/**
 * Definition for a binary tree node.
 **/
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private void dfs(TreeNode root, List<Integer> postOrder) {
        if (root == null)
            return;
        dfs(root.left, postOrder);
        dfs(root.right, postOrder);
        postOrder.add(root.val);
        return;
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        // N : no of Nodes
        // Time Complexity: O(2N)
        // Space Complexity: O(2N)
        // List<Integer> postOrder = new ArrayList<>();
        // Stack<TreeNode> stack = new Stack<>();
        // stack.push(root);
        // while (!stack.empty()){
        // TreeNode node = stack.pop();
        // if (node == null) continue;
        // postOrder.add(node.val);
        // stack.push(node.left);
        // stack.push(node.right);
        // }
        // return postOrder.reversed();

        // N : no of Nodes
        // Time Complexity: O(N)
        // Space Complexity: O(2N)
        List<Integer> postOrder = new ArrayList<>();
        dfs(root, postOrder);
        return postOrder;
    }
}